pkg/generator/github/github.go

```diff
@@ -0,0 +1,160 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package github
+
+import (
+	"context"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/golang-jwt/jwt/v5"
+	corev1 "k8s.io/api/core/v1"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	genv1alpha1 "github.com/external-secrets/external-secrets/apis/generators/v1alpha1"
+)
+
+type Generator struct {
+	httpClient *http.Client
+}
+
+type Github struct {
+	HTTP       *http.Client
+	Kube       client.Client
+	Namespace  string
+	URL        string
+	InstallTkn string
+}
+
+const (
+	defaultLoginUsername = `token`
+
+	errNoSpec    = "no config spec provided"
+	errParseSpec = "unable to parse spec: %w"
+	errGetToken  = "unable to get authorization token: %w"
+)
+
+func (g *Generator) Generate(ctx context.Context, jsonSpec *apiextensions.JSON, kube client.Client, namespace string) (map[string][]byte, error) {
+	return g.generate(
+		ctx,
+		jsonSpec,
+		kube,
+		namespace,
+	)
+}
+
+func (g *Generator) generate(
+	ctx context.Context,
+	jsonSpec *apiextensions.JSON,
+	kube client.Client,
+	namespace string) (map[string][]byte, error) {
+	if jsonSpec == nil {
+		return nil, fmt.Errorf(errNoSpec)
+	}
+	gh, err := newGHClient(ctx, kube, namespace, g.httpClient, jsonSpec)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+	// Github api expects POST request
+	req, err := http.NewRequestWithContext(context.TODO(), "POST", gh.URL, http.NoBody)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+	req.Header.Add("Authorization", "Bearer "+gh.InstallTkn)
+	req.Header.Add("Accept", "application/vnd.github.v3+json")
+
+	resp, err := gh.HTTP.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error performing request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// git access token
+	var gat map[string]interface{}
+	if err := json.NewDecoder(resp.Body).Decode(&gat); err != nil && resp.StatusCode < 300 {
+		return nil, fmt.Errorf("error decoding response: %w", err)
+	}
+
+	accessToken, ok := gat["token"].(string)
+	if !ok {
+		return nil, fmt.Errorf("token is not a string")
+	}
+	return map[string][]byte{
+		defaultLoginUsername: []byte(accessToken),
+	}, nil
+}
+
+func newGHClient(ctx context.Context, k client.Client, n string, hc *http.Client,
+	js *apiextensions.JSON) (*Github, error) {
+	if hc == nil {
+		hc = &http.Client{}
+	}
+	res, err := parseSpec(js.Raw)
+	if err != nil {
+		return nil, fmt.Errorf(errParseSpec, err)
+	}
+	gh := &Github{Kube: k, Namespace: n, HTTP: hc}
+
+	ghPath := fmt.Sprintf("/app/installations/%s/access_tokens", res.Spec.InstallID)
+	gh.URL = "https://api.github.com" + ghPath
```

Comment:
This URL should be configurable by the user since this can be enterprise or an on-prem installation.

Reply:
Wait a second... This URL should already be configured. Are you overwriting it here? 🤔

Reply:
url can be set by 
apis/generators/v1alpha1/generator_github.go:25
```
type GithubAccessTokenSpec struct {
	// URL configures the Github instance URL. Defaults to https://github.com/.
	URL       string `json:"url,omitempty"`
```

replaced with:
```
const (
	defaultGithubAPI     = "https://api.github.com"
) 
gh.URL = defaultGithubAPI + ghPath
if res.Spec.URL != "" {
	gh.URL = res.Spec.URL + ghPath
}
```